### PR TITLE
bazel: Use Xcode on macOS

### PR DESCRIPTION
**Pull request by fmeum**

### CODE REVIEW

**Comment by MarkusZoppelt** [comment]
Running `bazel build //:jazzer_release` still results in an error on my machine, if I do not use LLVM from homebrew.

```
[...]
ERROR: /private/var/tmp/_bazel_mz/6b6d52b1f712bd386568d1d984781c98/external/fmeum_rules_jni/jni/tools/libjvm_stub/BUILD.bazel:72:21: While resolving toolchains for target @fmeum_rules_jni//jni/tools/libjvm_stub:current_java_runtime: No matching toolchains found for types @bazel_tools//tools/cpp:toolchain_type.
To debug, rerun with --toolchain_resolution_debug='@bazel_tools//tools/cpp:toolchain_type'
If platforms or toolchains are a new concept for you, we'd encourage reading https://bazel.build/concepts/platforms-intro.
ERROR: Analysis of target '//:jazzer_release' failed; build aborted:
INFO: Elapsed time: 9.002s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (61 packages loaded, 665 targets configured)
    Fetching repository @remotejdk11_macos_aarch64; starting
    Fetching repository @local_config_cc; Running xcode-locator
    Fetching /private/var/tmp/_bazel_mz/6b6d52b1f712bd386568d1d984781c98/external/remotejdk11_macos_aarch64; Extracting zulu11.56.19-ca-jdk11.0.15-macosx_aarch64.tar.gz
    Fetching https://mirror.bazel.build/cdn.azul.com/zulu/bin/zulu11.56.19-ca-jdk11.0.15-macosx_x64.tar.gz
    Fetching https://raw.githubusercontent.com/openjdk/jdk/jdk-19%2B36/src/java.base/share/native/include/jni.h
    Fetching https://raw.githubusercontent.com/openjdk/jdk/jdk-19%2B36/src/java.base/unix/native/include/jni_md.h
```

**Comment by fmeum** [comment]
Could you rerun with `--toolchain_resolution_debug='@bazel_tools//tools/cpp:toolchain_type'` and post the logs? It looks as if Bazel doesn't find Xcode on your machine.

**Comment by 0xricksanchez** [comment]
Just to chime in `bazel build //:jazzer_release` works for me on this branch :)!

```shell
# [...]
INFO: Elapsed time: 78.820s, Critical Path: 21.89s
INFO: 701 processes: 132 internal, 500 darwin-sandbox, 8 local, 61 worker.
INFO: Build completed successfully, 701 total actions
$ ll bazel-bin/jazzer_release.tar.gz 
.r-xr-xr-x <user> <group> 8.4 MB Tue Dec 20 09:51:10 2022  bazel-bin/jazzer_release.tar.gz
```

**Comment by MarkusZoppelt** [comment]
> Could you rerun with `--toolchain_resolution_debug='@bazel_tools//tools/cpp:toolchain_type'` and post the logs? It looks as if Bazel doesn't find Xcode on your machine.

Here you go:
```
>  bazel build //:jazzer_release --toolchain_resolution_debug='@bazel_tools//tools/cpp:toolchain_type'
Starting local Bazel server and connecting to it...
INFO: ToolchainResolution:     Type @bazel_tools//tools/cpp:toolchain_type: target platform @local_config_platform//:host: Rejected toolchain @local_config_cc//:cc-compiler-armeabi-v7a; mismatching values: armv7, android
INFO: ToolchainResolution:     Type @bazel_tools//tools/cpp:toolchain_type: target platform //:macos_arm64: Rejected toolchain @local_config_cc//:cc-compiler-armeabi-v7a; mismatching values: armv7, android
INFO: ToolchainResolution:     Type @bazel_tools//tools/cpp:toolchain_type: target platform //:macos_x86_64: Rejected toolchain @local_config_cc//:cc-compiler-armeabi-v7a; mismatching values: armv7, android
INFO: ToolchainResolution:     Type @bazel_tools//tools/cpp:toolchain_type: target platform //:macos_arm64: Rejected toolchain @local_config_cc//:cc-compiler-armeabi-v7a; mismatching values: armv7, android
INFO: ToolchainResolution:     Type @bazel_tools//tools/cpp:toolchain_type: target platform //:macos_x86_64: Rejected toolchain @local_config_cc//:cc-compiler-darwin_arm64; mismatching values: arm64
INFO: ToolchainResolution:   Type @bazel_tools//tools/cpp:toolchain_type: target platform //:macos_arm64: execution @local_config_platform//:host: Selected toolchain @local_config_cc//:cc-compiler-darwin_arm64
INFO: ToolchainResolution:   Type @bazel_tools//tools/cpp:toolchain_type: target platform @local_config_platform//:host: execution @local_config_platform//:host: Selected toolchain @local_config_cc//:cc-compiler-darwin_arm64
INFO: ToolchainResolution:   Type @bazel_tools//tools/cpp:toolchain_type: target platform //:macos_x86_64: No toolchains found.
INFO: ToolchainResolution:   Type @bazel_tools//tools/cpp:toolchain_type: target platform //:macos_arm64: execution @local_config_platform//:host: Selected toolchain @local_config_cc//:cc-compiler-darwin_arm64
INFO: ToolchainResolution: Target platform //:macos_x86_64: Selected execution platform @local_config_platform//:host,
INFO: ToolchainResolution: Target platform //:macos_arm64: Selected execution platform @local_config_platform//:host, type @bazel_tools//tools/cpp:toolchain_type -> toolchain @local_config_cc//:cc-compiler-darwin_arm64
INFO: ToolchainResolution: Target platform @local_config_platform//:host: Selected execution platform @local_config_platform//:host, type @bazel_tools//tools/python:toolchain_type -> toolchain @bazel_tools//tools/python:_autodetecting_py_runtime_pair, type @bazel_tools//tools/cpp:toolchain_type -> toolchain @local_config_cc//:cc-compiler-darwin_arm64
INFO: ToolchainResolution: Target platform //:macos_arm64: Selected execution platform @local_config_platform//:host, type @bazel_tools//tools/cpp:toolchain_type -> toolchain @local_config_cc//:cc-compiler-darwin_arm64, type @bazel_tools//tools/jdk:toolchain_type -> toolchain //:java_non_prebuilt, type @bazel_tools//tools/jdk:runtime_toolchain_type -> toolchain @remotejdk11_macos_aarch64//:jdk
INFO: ToolchainResolution: Target platform //:macos_x86_64: Selected execution platform @local_config_platform//:host, type @bazel_tools//tools/jdk:toolchain_type -> toolchain //:java_non_prebuilt, type @bazel_tools//tools/jdk:runtime_toolchain_type -> toolchain @remotejdk11_macos//:jdk
INFO: ToolchainResolution:     Type @bazel_tools//tools/cpp:toolchain_type: target platform //:macos_x86_64: Rejected toolchain @local_config_cc//:cc-compiler-armeabi-v7a; mismatching values: armv7, android
INFO: ToolchainResolution:     Type @bazel_tools//tools/cpp:toolchain_type: target platform //:macos_x86_64: Rejected toolchain @local_config_cc//:cc-compiler-darwin_arm64; mismatching values: arm64
INFO: ToolchainResolution:   Type @bazel_tools//tools/cpp:toolchain_type: target platform //:macos_x86_64: No toolchains found.
INFO: Repository remotejdk11_macos instantiated at:
  /DEFAULT.WORKSPACE.SUFFIX:109:6: in <toplevel>
  /private/var/tmp/_bazel_mz/6b6d52b1f712bd386568d1d984781c98/external/bazel_tools/tools/build_defs/repo/utils.bzl:233:18: in maybe
  /private/var/tmp/_bazel_mz/6b6d52b1f712bd386568d1d984781c98/external/bazel_tools/tools/jdk/remote_java_repository.bzl:48:17: in remote_java_repository
Repository rule http_archive defined at:
  /private/var/tmp/_bazel_mz/6b6d52b1f712bd386568d1d984781c98/external/bazel_tools/tools/build_defs/repo/http.bzl:372:31: in <toplevel>
INFO: Repository com_github_openjdk_jdk_jni_h instantiated at:
  /Users/mz/code/jazzer/WORKSPACE.bazel:10:12: in <toplevel>
  /Users/mz/code/jazzer/init.bzl:30:27: in jazzer_init
  /private/var/tmp/_bazel_mz/6b6d52b1f712bd386568d1d984781c98/external/fmeum_rules_jni/jni/internal/repositories.bzl:39:10: in rules_jni_dependencies
  /private/var/tmp/_bazel_mz/6b6d52b1f712bd386568d1d984781c98/external/bazel_tools/tools/build_defs/repo/utils.bzl:233:18: in maybe
Repository rule http_file defined at:
  /private/var/tmp/_bazel_mz/6b6d52b1f712bd386568d1d984781c98/external/bazel_tools/tools/build_defs/repo/http.bzl:466:28: in <toplevel>
INFO: Repository com_github_openjdk_jdk_unix_jni_md_h instantiated at:
  /Users/mz/code/jazzer/WORKSPACE.bazel:10:12: in <toplevel>
  /Users/mz/code/jazzer/init.bzl:30:27: in jazzer_init
  /private/var/tmp/_bazel_mz/6b6d52b1f712bd386568d1d984781c98/external/fmeum_rules_jni/jni/internal/repositories.bzl:46:10: in rules_jni_dependencies
  /private/var/tmp/_bazel_mz/6b6d52b1f712bd386568d1d984781c98/external/bazel_tools/tools/build_defs/repo/utils.bzl:233:18: in maybe
Repository rule http_file defined at:
  /private/var/tmp/_bazel_mz/6b6d52b1f712bd386568d1d984781c98/external/bazel_tools/tools/build_defs/repo/http.bzl:466:28: in <toplevel>
INFO: Repository remotejdk11_macos_aarch64 instantiated at:
  /DEFAULT.WORKSPACE.SUFFIX:125:6: in <toplevel>
  /private/var/tmp/_bazel_mz/6b6d52b1f712bd386568d1d984781c98/external/bazel_tools/tools/build_defs/repo/utils.bzl:233:18: in maybe
  /private/var/tmp/_bazel_mz/6b6d52b1f712bd386568d1d984781c98/external/bazel_tools/tools/jdk/remote_java_repository.bzl:48:17: in remote_java_repository
Repository rule http_archive defined at:
  /private/var/tmp/_bazel_mz/6b6d52b1f712bd386568d1d984781c98/external/bazel_tools/tools/build_defs/repo/http.bzl:372:31: in <toplevel>
ERROR: /private/var/tmp/_bazel_mz/6b6d52b1f712bd386568d1d984781c98/external/fmeum_rules_jni/jni/tools/libjvm_stub/BUILD.bazel:72:21: While resolving toolchains for target @fmeum_rules_jni//jni/tools/libjvm_stub:current_java_runtime: No matching toolchains found for types @bazel_tools//tools/cpp:toolchain_type.
To debug, rerun with --toolchain_resolution_debug='@bazel_tools//tools/cpp:toolchain_type'
If platforms or toolchains are a new concept for you, we'd encourage reading https://bazel.build/concepts/platforms-intro.
ERROR: Analysis of target '//:jazzer_release' failed; build aborted:
INFO: Elapsed time: 1.901s
INFO: 0 processes.
FAILED: Build did NOT complete successfully (60 packages loaded, 663 targets configured)
    Fetching repository @local_config_xcode; Building xcode-locator
    Fetching repository @local_config_cc; Building xcode-locator
```

fwiw, I don't have Xcode.app installed. Only the developer tools installed by `xcode-select --install`

**Comment by fmeum** [comment]
@MarkusZoppelt The logs indicate that Xcode wasn't found. I don't know whether Bazel should work with just the developer tools - I will try to get some clarity on that.

**Comment by fmeum** [comment]
Bazel requires Xcode itself, not just the developer tools: https://bazelbuild.slack.com/archives/CD3QY5C2X/p1671572167312109?thread_ts=1671572167.312109&cid=CD3QY5C2X

**Comment by MarkusZoppelt** [comment]
> Bazel requires Xcode itself, not just the developer tools: https://bazelbuild.slack.com/archives/CD3QY5C2X/p1671572167312109?thread_ts=1671572167.312109&cid=CD3QY5C2X

Alright. That's fine imho. Maybe we should add that also to the documentation in case other people have the same configuration.

File: bazel/compat.bzl [hunk 1]
```diff
@@ -26,3 +26,11 @@ LINUX_ONLY = select({
     "@platforms//os:linux": [],
     "//conditions:default": ["@platforms//:incompatible"],
 })
+
+MACOS_MULTI_PLATFORM = select({
```
Comment (bertschneider) [inline bazel/compat.bzl:30]: This enables multi platform builds for macOS but is also evaluated for Windows and Linux, right? If so, should we remove the `MACOS` prefix?

File: launcher/BUILD.bazel [hunk 2]
```diff
@@ -24,12 +26,18 @@ cc_library(
...
 cc_binary(
-    name = "jazzer",
+    name = "jazzer_single_arch",
     linkstatic = True,
-    visibility = ["//visibility:public"],
+    tags = ["manual"],
     deps = [":jazzer_main"],
 )
 
+universal_binary(
```
Comment (bertschneider) [inline launcher/BUILD.bazel:35]: Could you add a comment regarding non-macOS platforms? I was a somewhat confused and a comment could help the next reader along. The docs state that `universal_binary` would just create a symlink to the input binary.

File: .github/workflows/run-all-tests.yml [hunk 3]
```diff
@@ -60,17 +60,6 @@ jobs:
         run: .github/scripts/echoBuildBuddyConfig.sh ${{ secrets.BUILDBUDDY_API_KEY }} >> $GITHUB_ENV
         shell: bash
 
-      - name: Set CC to LLVM clang on macOS
```
Comment (bertschneider) [inline .github/workflows/run-all-tests.yml:63]: Is the macOS restriction in the matrix section further up still needed?

**Review by bertschneider — APPROVED** [review]

File: bazel/compat.bzl (hunk 1 above)
Comment (fmeum) [reply bazel/compat.bzl]: Good point. I renamed it to `MULTI_PLATFORM_ON_MACOS` to make it clear that it only has an effect on macOS but can still be used on every OS.

File: .github/workflows/run-all-tests.yml (hunk 3 above)
Comment (fmeum) [reply .github/workflows/run-all-tests.yml:63]: Thanks, I don't think so. Removed it.

File: launcher/BUILD.bazel (hunk 2 above)
Comment (fmeum) [reply launcher/BUILD.bazel:37]: Added a comment.

File: bazel/compat.bzl (hunk 1 above)
Comment (fmeum) [reply bazel/compat.bzl]: Reconsidered and removed the suffix altogether as you suggested - we may want to add Windows ARM support with this mechanism.

**Review by bertschneider — APPROVED** [review]
Great simplification!

**Pull request merged**